src/services/programs.js

```diff
@@ -623,6 +625,349 @@ module.exports = class ProgramsHelper {
+				},
+				resource_id: {
+					[Op.in]: [programId, ...resourceIds],
+				},
+				status: common.COMMENT_STATUS_OPEN,
+			})
+
+			if (comments.count > 0) {
+				return responses.failureResponse({
+					message: 'ALL_COMMENTS_NOT_RESOLVED',
+					statusCode: httpStatusCode.bad_request,
+					responseCode: 'CLIENT_ERROR',
+				})
+			}
+
+			// Check that the note character limit does not exceed the maximum limit
+			if (bodyData?.notes?.length > process.env.MAX_RESOURCE_NOTE_LENGTH) {
+				return responses.failureResponse({
+					message: 'RESOURCE_NOTE_LENGTH_EXCEEDED',
+					statusCode: httpStatusCode.bad_request,
+					responseCode: 'CLIENT_ERROR',
+				})
+			}
+
+			//validate reviewers
+			let reviewerIds = []
+			if (bodyData.reviewer_ids && bodyData.reviewer_ids.length > 0) {
+				const uniqueReviewerIds = utils.getUniqueElements(bodyData.reviewer_ids)
+				const reviewers = await userRequests.list(common.REVIEWER, '', '', '', userDetails.organization_id, {
+					user_ids: uniqueReviewerIds,
+					excluded_user_ids: [userDetails.id],
+				})
+
+				if (!reviewers.success) throw new Error('REVIEWER_IDS_NOT_FOUND')
+
+				//written as a backup will remove once the user service PR merged
+				if (Array.isArray(reviewers?.data?.result?.data) && reviewers.data.result.data.length > 0) {
+					reviewerIds = reviewers.data.result.data.map((item) => item.id)
+				} else {
+					// If no valid reviewers data is found, return an error response
+					throw new Error('REVIEWER_IDS_NOT_FOUND')
+				}
+
+				//return error message if the reviewer is invalid or not found
+				if (uniqueReviewerIds.length > reviewers.data.result.data.length) {
+					throw new Error('REVIEWER_IDS_NOT_FOUND')
+				}
+			}
+
+			// Check if any resources are added to program
+			if (resourceIds.length == 0) {
+				validationErrors.push(
+					utils.errorObject(common.RESOURCE_TYPE_PROGRAM, 'resources', `Atleast one resource is mandatory.`)
+				)
+			}
+
+			//get all entity type validations for program
+			let entityTypes = await entityModelMappingQuery.findEntityTypesAndEntities(
+				{
+					model: common.RESOURCE_TYPE_PROGRAM,
+					status: common.STATUS_ACTIVE,
+				},
+				programData.organization_id,
+				['id', 'value', 'has_entities', 'validations']
+			)
+
+			let basePath = ''
+			//validate program data
+			const programValidationPromises = entityTypes.map((entityType) => {
+				validateEntityData(programData, entityType, common.RESOURCE_TYPE_PROGRAM, basePath, validationErrors)
+			})
+
+			await Promise.all(programValidationPromises)
+
+			let resourceValidationErrors = []
+
+			let resourceEntityTypes = await entityModelMappingQuery.findEntityTypesAndEntities(
+				{
+					model: {
+						[Op.in]: resourceTypes,
+					},
+					status: common.STATUS_ACTIVE,
+				},
+				programData.organization_id,
+				['id', 'value', 'has_entities', 'validations']
+			)
+
+			const resourcesValidationPromise = resourceData.map(async (resource, index) => {
```

Comment:
@adithyadinesh0412  can we add targeting criteria and start_date end_check in a separate common function? like checking end date and targeting some of the check is common in both